aspnetcore/blazor/tutorials/movie-database-app/part-6.md

```diff
@@ -26,66 +26,45 @@ This part of the tutorial series covers adding a search feature to the movies `I
 The [`QuickGrid`](xref:Microsoft.AspNetCore.Components.QuickGrid) component is used by the movie `Index` component (`Components/MoviePages/Index.razor`) to display movies from the database:
 
 ```razor
-<QuickGrid Class="table" Items="DB.Movie">
+<QuickGrid Class="table" Items="DbFactory.CreateDbContext().Movie">
     ...
 </QuickGrid>
 ```
 
-The <xref:Microsoft.AspNetCore.Components.QuickGrid.QuickGrid%601.Items%2A> parameter receives a nullable `IQueryable<TGridItem>`, where `TGridItem` is the type of data represented by each row in the grid. In this case, the `TGridItem` is a `DbSet<Movie>` obtained from the injected `BlazorWebAppMoviesContext` (`DB`).
+The <xref:Microsoft.AspNetCore.Components.QuickGrid.QuickGrid%601.Items%2A> parameter receives an `IQueryable<TGridItem>`, where `TGridItem` is the type of data represented by each row in the grid (`Movie`). <xref:Microsoft.AspNetCore.Components.QuickGrid.QuickGrid%601.Items%2A> is assigned a collection of movie entities (`DbSet<Movie>`) obtained from the created database context (`CreateDbContext`) of the injected database context factory (`DbFactory`).
 
 To make the `QuickGrid` component filter on the movie title, the `Index` component should:
 
 * Set a filter string as a *component parameter* from the query string.
 * If the parameter has a value, filter the movies returned from the database.
 * Provide an input for the user to provide the filter string and a button to trigger a reload using the filter.
 
-Start by adding the following `@code` block of C# code to the `Index` component:
+Start by adding the following `@code` block of C# code to the `Index` component (`MoviePages/Index.razor`):
 
 ```razor
 @code {
-    private IQueryable<Movie>? movies;
-
     [SupplyParameterFromQuery]
     public string? TitleFilter { get; set; }
-
-    protected override void OnParametersSet()
-    {
-        if (!string.IsNullOrEmpty(TitleFilter))
-        {
-            movies = DB.Movie.Where(
-                s => !string.IsNullOrEmpty(s.Title) ? 
-                    s.Title.Contains(TitleFilter) : false);
-        }
-        else
-        {
-            movies = DB.Movie;
-        }
-    }
+    
+    private IQueryable<Movie> FilteredMovies => DbFactory.CreateDbContext().Movie
```

Comment:
Here's another case where we probably need to handle disposing the DbContext.

Reply:
IIRC 🤔, Jeremy's code that disposes only does so when it's assigned. I may have guessed incorrectly that in these two cases (here and for the `Items` assignment) that they would be OK like this.

Reply:
Adding folks to review and comment on whether we need to dispose the created DbContext here as well: @JeremyLikness @javiercn @deepchoudhery @halter73

Reply:
I imagine we should be disposing instances per component ala [this](https://learn.microsoft.com/en-us/aspnet/core/blazor/blazor-ef-core?view=aspnetcore-8.0#scope-to-the-component-lifetime:~:text=component%20is%20disposed%3A-,C%23,public%20void%20Dispose()%0A%7B%0A%20%20%20%20Context%3F.Dispose()%3B%0A%7D,-Finally%2C%20OnInitializedAsync) example

Reply:
@deepchoudhery ... AFAIK, Jeremy only disposes if the context is assigned. In both of the cases here, no assignment is made.

Reply:
we should dispose - page has hooks for shared context scoped to page and components should dispose in their lifetime. I haven't reviewed the entire PR, so I don't have all the context (pardon the pun) but I would at least hook into one of the page lifecycle overrides to grab the movie list once and dispose the context.

Reply:
Gotcha @JeremyLikness ... I can take care of that. I'll get it squared away first thing tomorrow morning.

Reply:
@JeremyLikness and I chatted about this some more. We should dispose, but we think we should stick with the pattern where we create-use-dispose the context for each unit of work instead of tying the context lifetime to the lifetime of the component.

Reply:
BTW ... Based on the discussion, this guidance isn't correct ...

https://learn.microsoft.com/en-us/aspnet/core/blazor/components/quickgrid?view=aspnetcore-8.0&tabs=visual-studio#entity-framework-core-ef-core-data-source

Reply:
I opened https://github.com/dotnet/Scaffolding/issues/2870 to address the underlying issue in the Blazor CRUD scaffolder.

Reply:
@danroth27 ... Questions ...

* Property or field for the context?
* Would you like that init and disposal set up async?

```razor
@page "/movies"
@rendermode InteractiveServer
@using BlazorWebAppMovies.Data
@using Microsoft.EntityFrameworkCore
@using Microsoft.AspNetCore.Components.QuickGrid
@using BlazorWebAppMovies.Models
@inject IDbContextFactory<BlazorWebAppMovies.Data.BlazorWebAppMoviesContext> DbFactory
@implements IAsyncDisposable

<PageTitle>Index</PageTitle>

<h1>Index</h1>

<p>
    <input type="search" @bind="titleFilter" @bind:event="oninput" />
</p>

<p>
    <a href="movies/create">Create New</a>
</p>

<QuickGrid Class="table" Items="FilteredMovies" Pagination="pagination">
    <PropertyColumn Property="movie => movie.Title" Sortable="true" />
    <PropertyColumn Property="movie => movie.ReleaseDate" Title="Release Date" />
    <PropertyColumn Property="movie => movie.Genre" />
    <PropertyColumn Property="movie => movie.Price" />
    <PropertyColumn Property="movie => movie.Rating" />

    <TemplateColumn Context="movie">
        <a href="@($"movies/edit?id={movie.Id}")">Edit</a> |
        <a href="@($"movies/details?id={movie.Id}")">Details</a> |
        <a href="@($"movies/delete?id={movie.Id}")">Delete</a>
    </TemplateColumn>
</QuickGrid>

<Paginator State="pagination" />

@code {
    PaginationState pagination = new PaginationState { ItemsPerPage = 10 };
    private string titleFilter = string.Empty;
    private BlazorWebAppMoviesContext context = default!;
    
    protected override async Task OnInitializedAsync()
    {
        context = await DbFactory.CreateDbContextAsync();
    }

    private IQueryable<Movie> FilteredMovies => 
        context.Movie.Where(m => m.Title!.Contains(titleFilter));

    public async ValueTask DisposeAsync() => await context.DisposeAsync();
}

```

Reply:
@JeremyLikness ... This component has to filter entities with an `IQueryable` for the `QuickGrid` to return entities based on the `DbContext`, as opposed to getting the list on load. The above code works :tada:, but smells 💩👃 exist if using field is a bad idea and/or if async init/disposal are bad ideas.

Reply:
Per our chat on the scaffolding issue: The `Task` not completing in time for filling the grid will result in my stripping the async out (probably). I wish binding the items in some simple fashion would ✨ *automagically* ✨ fill the grid when the `Task` completes. I'll wait for more info, but now the code is headed toward ...

```razor
@implements IDisposable

...

protected override void OnInitialized()
{
        context = DbFactory.CreateDbContextAsync();
}

...

public Dispose() => context?.Dispose();
```